src/libasr/asr_utils.h

```diff
@@ -1144,8 +1144,19 @@ static inline std::string binop_to_str_python(const ASR::binopType t) {
 }
 
 static inline bool is_immutable(const ASR::ttype_t *type) {
+    bool is_integer = ASR::is_a<ASR::Integer_t>(*type);
+    bool is_dimensionless = false;
+    bool is_real = ASR::is_a<ASR::Real_t>(*type);
+    if(is_integer){
+        ASR::Integer_t *integer = ASR::down_cast<ASR::Integer_t>(type);
+        is_dimensionless = integer->n_dims == 0;
+    }
+    if(is_real){
+        ASR::Real_t *real = ASR::down_cast<ASR::Real_t>(type);
+        is_dimensionless = real->n_dims == 0;
+    }
     return ((ASR::is_a<ASR::Character_t>(*type) || ASR::is_a<ASR::Tuple_t>(*type)
-        || ASR::is_a<ASR::Complex_t>(*type)));
+        || ASR::is_a<ASR::Complex_t>(*type)) || is_dimensionless);
```

Comment:
```suggestion
        || ASR::is_a<ASR::Complex_t>(*type)) || !ASRUtils::is_array(type));
```

Reply:
Didn't realize there was a function for this :) I'll try out this suggestion. Thanks.

Reply:
I tried this diff directly but it doesn't seem to work -
```
const LCompilers::ASR::ttype_t *type
argument of type "const LCompilers::ASR::ttype_t *" is incompatible with parameter of type "LCompilers::ASR::ttype_t *"C/C++(167)
```
I will do some more investigation into this.

Reply:
@Thirumalai-Shaktivel @certik This is the only roadblock to this pull request, rest the results are as expected. Could you have a look at it. Let's get this in!

Reply:
Pass `type` using an `explicit` cast: `!ASRUtils::is_array((ASR::ttype_t *) type)`